clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1060,238 +1061,348 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
-  const ArrayType *ArrayTy = Ctx.getAsArrayType(FAMDecl->getType());
-  CharUnits Size = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
-  llvm::Constant *ElemSize =
-      llvm::ConstantInt::get(ResType, Size.getQuantity(), IsSigned);
-  Value *Res =
-      Builder.CreateMul(CountedByInst, ElemSize, "", !IsSigned, IsSigned);
-  Res = Builder.CreateIntCast(Res, ResType, IsSigned);
-
-  // A negative \p IdxInst or \p CountedByInst means that the index lands
-  // outside of the flexible array member. If that's the case, we want to
-  // return 0.
-  Value *Cmp = Builder.CreateIsNotNeg(CountedByInst);
-  if (IdxInst)
-    Cmp = Builder.CreateAnd(Builder.CreateIsNotNeg(IdxInst), Cmp);
-
-  return Builder.CreateSelect(Cmp, Res, ConstantInt::get(ResType, 0, IsSigned));
+    Index = EmitScalarExpr(Idx);
+    Index = Builder.CreateIntCast(Index, ResType, IdxSigned, "index");
+  }
+
+  //  size_t flexible_array_member_base_size = sizeof (*ptr->array);
+  const ArrayType *ArrayTy = Ctx.getAsArrayType(FlexibleArrayMemberTy);
+  CharUnits BaseSize = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
+  auto *FlexibleArrayMemberBaseSize =
+      llvm::ConstantInt::get(ResType, BaseSize.getQuantity(), IsSigned);
+
+  //  size_t flexible_array_member_size =
+  //          count * flexible_array_member_base_size;
+  Value *FlexibleArrayMemberSize =
+      Builder.CreateMul(Count, FlexibleArrayMemberBaseSize,
+                        "flexible_array_member_size", !IsSigned, IsSigned);
+
+  auto CheckForNegative = [&](Value *Res) {
+    Value *Cmp = Builder.CreateIsNotNeg(Res);
+    if (Idx)
+      Cmp = Builder.CreateAnd(Builder.CreateIsNotNeg(Index), Cmp);
+    return Builder.CreateSelect(Cmp, Res,
+                                ConstantInt::get(ResType, 0, IsSigned));
+  };
+
+  if (FlexibleArrayMemberFD == FD) {
+    if (Idx) { // Option (2) '&ptr->array[idx]'
+      //  size_t index_size = index * flexible_array_member_base_size;
+      Value *IndexSize = Builder.CreateMul(FlexibleArrayMemberBaseSize, Index,
+                                           "index_size", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size - index_size;
+      return CheckForNegative(Builder.CreateSub(
+          FlexibleArrayMemberSize, IndexSize, "result", !IsSigned, IsSigned));
+    } else { // Option (1) 'ptr->array'
+      //  return flexible_array_member_size;
+      return CheckForNegative(FlexibleArrayMemberSize);
+    }
+  } else {
+    //  size_t sizeof_struct = sizeof (struct s);
+    llvm::StructType *StructTy = getTypes().getCGRecordLayout(RD).getLLVMType();
+    const llvm::DataLayout &Layout = CGM.getDataLayout();
+    TypeSize Size = Layout.getTypeSizeInBits(StructTy);
+    Value *SizeofStruct =
+        llvm::ConstantInt::get(ResType, Size.getKnownMinValue() / CharWidth);
+
+    //  size_t field_offset = offsetof (struct s, field);
+    int64_t Offset = 0;
+    GetFieldOffset(Ctx, RD, FD, Offset);
```

Comment:
Do we need to check the return value from this call to `GetFieldOffset`? If so, consider marking `GetFieldOffset` as `[[nodiscard]]`.  Otherwise, I wonder if we can make `GetFieldOffset` infallible?  Maybe returning an optional int is better than having a return code which isn't checked (and in the case it's false, is `Offset` valid to use?).

If it fails, at this point we've already had builder emit a bunch of instructions...

Reply:
`GetFieldOffset's` return value is meant for the recursive call, and it shouldn't fail, but I'll change it around to catch any potential issues.